test/functional/test_framework/test_framework.py

```diff
@@ -307,11 +307,12 @@ def shutdown(self):
                 node.cleanup_on_exit = False
             self.log.info("Note: bitcoinds were not stopped and may still be running")
 
+        perf_available = any(bool(node.perf_available) for node in self.nodes)
         should_clean_up = (
             not self.options.nocleanup and
             not self.options.noshutdown and
             self.success != TestStatus.FAILED and
-            not self.options.perf
+            not (self.options.perf and perf_available)
```

Comment:
When `perf_available` is true, `self.options.perf` is not always true.

By checking only for `perf_available` (which is set only if perf is actually active) you are also fixing an issue where if [`profile_with_perf`](https://github.com/bitcoin/bitcoin/tree/master/test/functional#benchmarking-with-perf) is used, the directory still gets cleaned up, and therefore the perf data are lost.
```suggestion
            not (perf_available)
```
and also should check for `perf_available` instead of `self.options.perf` as part of the below conditional statement.

Reply:
Make sense, gonna address it!